File: app/src/main/java/com/faendir/lightning_launcher/multitool/calendar/CalendarPreference.java
package com.faendir.lightning_launcher.multitool.calendar;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.res.TypedArray;
import android.database.Cursor;
import android.provider.CalendarContract.Calendars;
import android.text.TextUtils;
import android.util.AttributeSet;
import androidx.core.content.ContextCompat;
import androidx.preference.MultiSelectListPreference;
import com.faendir.lightning_launcher.multitool.settings.SummaryPreference;
import java9.util.stream.Collectors;
import java9.util.stream.StreamSupport;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * @author lukas
 * @since 10.08.18
 */
public class CalendarPreference extends MultiSelectListPreference implements SummaryPreference {
    public CalendarPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        refresh();
    }

    private List<CharSequence> getSelectedEntries() {
        List<CharSequence> entries = Arrays.asList(getEntries());
        List<CharSequence> values = Arrays.asList(getEntryValues());
        return StreamSupport.stream(getValues()).map(values::indexOf).filter(index -> index >= 0).map(entries::get).sorted().collect(Collectors.toList());
    }

    @Override
    public CharSequence getSummaryText() {
        return TextUtils.join(", ", getSelectedEntries());
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        Set<String> calendars = new HashSet<>();
        if (ContextCompat.checkSelfPermission(getContext(), Manifest.permission.READ_CALENDAR) == PackageManager.PERMISSION_GRANTED) {
            try (Cursor cursor = getContext().getContentResolver().query(Calendars.CONTENT_URI, new String[]{Calendars._ID}, Calendars.VISIBLE + " = 1", null, Calendars._ID + " ASC")) {
                if (cursor != null) {
                    while (cursor.moveToNext()) {
                        calendars.add(cursor.getString(0));
                    }
                }
            }
        }
        return calendars;
    }

    public void refresh() {
        Map<String, String> calendars = new HashMap<>();
        if (ContextCompat.checkSelfPermission(getContext(), Manifest.permission.READ_CALENDAR) == PackageManager.PERMISSION_GRANTED) {
            try (Cursor cursor = getContext().getContentResolver()
                    .query(Calendars.CONTENT_URI, new String[]{Calendars._ID, Calendars.NAME}, Calendars.VISIBLE + " = 1", null, Calendars._ID + " ASC")) {
                if (cursor != null) {
                    while (cursor.moveToNext()) {
                        calendars.put(cursor.getString(0), cursor.getString(1));
                    }
                }
            }
        }
        setEntryValues(calendars.keySet().toArray(new CharSequence[0]));
        setEntries(calendars.values().toArray(new CharSequence[0]));
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/calendar/CalendarPreference.kt
package com.faendir.lightning_launcher.multitool.calendar

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.content.res.TypedArray
import android.provider.CalendarContract.Calendars
import android.text.TextUtils
import android.util.AttributeSet
import androidx.core.content.ContextCompat
import androidx.preference.MultiSelectListPreference
import com.faendir.lightning_launcher.multitool.settings.SummaryPreference
import java.util.*

/**
 * @author lukas
 * @since 10.08.18
 */
class CalendarPreference(context: Context, attrs: AttributeSet) : MultiSelectListPreference(context, attrs), SummaryPreference {

    private fun getSelectedEntries(): List<String> {
            val entries = entries.map { it.toString() }.toList()
            val values = entryValues.map { it.toString() }.toList()
            return getValues().asSequence().map { values.indexOf(it) }.filter { it >= 0 }.map { entries[it] }.sorted().toList()
        }

    init {
        refresh()
    }

    override fun getSummaryText(): CharSequence {
        return TextUtils.join(", ", getSelectedEntries())
    }

    override fun onGetDefaultValue(a: TypedArray, index: Int): Any {
        val calendars = HashSet<String>()
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_CALENDAR) == PackageManager.PERMISSION_GRANTED) {
            context.contentResolver.query(Calendars.CONTENT_URI, arrayOf(Calendars._ID), Calendars.VISIBLE + " = 1", null, Calendars._ID + " ASC")?.use {
                while (it.moveToNext()) {
                    calendars.add(it.getString(0))
                }
            }
        }
        return calendars
    }

    fun refresh() {
        val calendars = HashMap<String, String>()
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_CALENDAR) == PackageManager.PERMISSION_GRANTED) {
            context.contentResolver
                    .query(Calendars.CONTENT_URI, arrayOf(Calendars._ID, Calendars.NAME), Calendars.VISIBLE + " = 1", null, Calendars._ID + " ASC")?.use {
                        while (it.moveToNext()) {
                            calendars[it.getString(0)] = it.getString(1)
                        }
                    }
        }
        entryValues = calendars.keys.toTypedArray<CharSequence>()
        entries = calendars.values.toTypedArray<CharSequence>()
    }
}
